### PR TITLE
[FLINK-24897] Enable Application/Per-job mode on YARN to use usrlib

### PR DESCRIPTION
**Pull request by bgeng777**
## What is the purpose of the change

This pull request aims at enabling application/per-job mode on YARN to use usrlib easier. 

## Brief change log
  - Make `usrlib`(if exists) under FLINK_HOME be uploaded to the cluster automatically for both Application and Per-job mode on YARN.  If we add `usrlib`  in ship files again or specify `usrlib` in `yarn.provided.lib.dirs`, exception will be thrown. 
  - Jars in `usrlib` will be loaded by user classloader only when `yarn.per-job-cluster.include-user-jar` is DISABLED. 

## Verifying this change

This change added tests and can be verified as follows:
  - Add tests that validate the shipping logic
  - Manually verified the classloader is chosen correctly when `yarn.per-job-cluster.include-user-jar` is DISABLED

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no
  - 
## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs(but need improvement)

### CODE REVIEW

**Comment by bgeng777** [comment]
@flinkbot run azure

**Comment by bgeng777** [comment]
@flinkbot run azure

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java [hunk 1]
```diff
@@ -438,6 +446,18 @@ private static boolean isPlugin(Path path) {
         return false;
     }
 
+    private static boolean isUsrLib(Path path) {
```
Comment (wangyang0918) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java:449]: I suggest to introduce the `isUsrLibDirIncludedInProvidedLib` instead and then `checkArgument` in the constructor of `YarnApplicationFileUploader`. Because the files in the provided lib will be relativized, so we just need to check the name of provided lib.

```
    private boolean isUsrLibDirIncludedInProvidedLib(final List<Path> providedLibDirs)
            throws IOException {
        for (Path path : providedLibDirs) {
            final FileStatus fileStatus = fileSystem.getFileStatus(path);
            if (fileStatus.isDirectory()
                    && ConfigConstants.DEFAULT_FLINK_USR_LIB_DIR.equals(path.getName())) {
                return true;
            }
        }
        return false;
    }
```

**Review by wangyang0918 — CHANGES_REQUESTED** [review]
@bgeng777 Thanks for creating this PR. I think your proposal makes sense and could work. And I have left some comments about the implementation. Please have a look.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 2]
```diff
@@ -1685,6 +1699,35 @@ void addLibFoldersToShipFiles(Collection<File> effectiveShipFiles) {
...
+            Collection<File> effectiveShipFiles, Collection<File> systemShipFiles) {
+        // Add usrlib folder to the ship files if it exists
+        // Classes in the folder will be loaded by UserClassLoader if CLASSPATH_INCLUDE_USER_JAR is
+        // DISABLED.
+        final Optional<File> usrLibDir = getLocalUsrLibDirectory();
+
+        if (usrLibDir.isPresent()) {
+            File usrLibDirFile = usrLibDir.get();
+            if (usrLibDirFile.isDirectory()) {
+                checkArgument(
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1713]: Hi  Yang, thanks for the feedback. In fact, my first version of codes is pretty like your code snippet. 
But after thinking more carefully, I use my current implementation (including writing my own getLocalUsrLibDirectory()) for the following reason:
1. AFAIK, YarnClusterDescriptor is executed in the flink client side while ClusterEntryPointUtil is executed on the remote cluster side. If I use the `ClusterEntrypointUtils.tryFindUserLibDirectory()` in client side, due to its impl, the user must specify the FLINK_LIB_DIR correctly. But IIUC, we do not have such requirement before when using YARN. 
2. In my `getLocalUsrLibDirectory()`, I use the FLINK_HOME as a pivot to find the `usrlib`. This is not so perfect but comparing with FLINK_LIB_DIR, FLINK_HOME may be more widely used.

In summary, I bet the key question is if we should ask the action of finding `usrlib` to adopt same logic(using FLINK_LIB_DIR or use current working dir) in both client and cluster side.  Do you have any suggestion?

File: flink-yarn/pom.xml [hunk 3]
```diff
@@ -130,6 +130,11 @@ under the License.
 				</exclusion>
 			</exclusions>
 		</dependency>
+		<dependency>
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/pom.xml:133]: My bad. Removed.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 4]
```diff
@@ -910,14 +910,28 @@ private ApplicationReport startAppMaster(
         }
 
         // Upload and register user jars
-        final List<String> userClassPaths =
+        List<String> userClassPaths =
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:913]: Because we need to check if usrlib exists later. If usrlib exists, we need `userClassPaths.addAll(usrLibClassPaths);` so we can reuse current code to make other options o YarnConfigOptions.UserJarInclusion work as expected.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 5]
```diff
@@ -1799,4 +1842,27 @@ public static void logDetachedClusterInformation(
                 yarnApplicationId,
                 yarnApplicationId);
     }
+
+    /**
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1846]: See https://github.com/apache/flink/pull/18531/files/5eb98df7422ffca67eb4d258189ff1eadb36731e#discussion_r801320099

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java [hunk 6]
```diff
@@ -335,7 +335,7 @@ public YarnLocalResourceDescriptor uploadFlinkDist(final Path localJarPath)
      *
      * @return list of class paths with the file name
      */
-    List<String> registerProvidedLocalResources() {
+    List<String> registerProvidedLocalResources(boolean shouldUsrLibExistInRemote) {
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java:338]: Yes, I totally agree with the assumption and that it just have the same semantic with ship files. 
But in ship files, we will check if `usrlib` exists in ship files and if so and `INCLUDE_USER_JAR` is DISABLED, we will throw exception due to our previous discussion. 
The reason for me to adding such code is for this case: users have local `usrlib` and in the remote dirs specified by `provided.lib.dirs`, there is also a remote `usrlib`. I want to get rid of  such case as it could lead to overriding files in YARN cache's `usrlib`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1713]: Except for the naming, I do not think the `ClusterEntrypointUtils.tryFindUserLibDirectory()` should only work in the cluster side. If a user configure a customized `FLINK_LIB_DIR`, then he/she should also have the `usrlib` in such location. I think it is reasonable.

For the naming, maybe we could have the `UserLibUtils.java`, which contains all the utility methods for `usrlib`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:913]: Even though we could still have the `final`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java:338]: I think no matter the `INCLUDE_USER_JAR` is configured, we should always not include `usrlib` in the user shipped files. This is also true for the provided lib.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 7]
```diff
@@ -872,14 +872,14 @@ private ApplicationReport startAppMaster(
 
             jobGraph.writeUserArtifactEntriesToConfiguration();
         }
-
+        boolean isLocalUsrLibExists = getLocalUsrLibDirectory().isPresent();
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Fixed in ef3f0104fc31def407b74b4698d5f7dd0404a333.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java (hunk 1 above)
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java]: Make sense. Fixed in ef3f0104fc31def407b74b4698d5f7dd0404a333

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1712]: Agree with the naming and the `FLINK_LIB_DIR` clarification.
Just one follow-up: I know that `FLINK_LIB_DIR` will be set by `config.sh` if it is empty when we run `flink run ....`.
But when YARN run commands like `/bin/bash -c /usr/lib/jvm/java-1.8.0/bin/java -Xmx1073741824 -Xms1073741824 -XX:MaxMetaspaceSize=268435456 -Dlog.file=/mnt/disk1/log/hadoop-yarn/containers/application_1643263086460_0041/container_1643263086460_0041_01_000001/jobmanager.log -Dlog4j.configuration=file:log4j.properties -Dlog4j.configurationFile=file:log4j.properties org.apache.flink.yarn.entrypoint.YarnJobClusterEntrypoint -D jobmanager.memory.off-heap.size=134217728b -D jobmanager.memory.jvm-overhead.min=201326592b -D jobmanager.memory.jvm-metaspace.size=268435456b -D jobmanager.memory.heap.size=1073741824b -D jobmanager.memory.jvm-overhead.max=201326592b 1> /mnt/disk1/log/hadoop-yarn/containers/application_1643263086460_0041/container_1643263086460_0041_01_000001/jobmanager.out 2> /mnt/disk1/log/hadoop-yarn/containers/application_1643263086460_0041/container_1643263086460_0041_01_000001/jobmanager.err`, I think the `bin/config.sh` will not executed? If I am wrong, please correct me. Thanks!

[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1712]: What I mean is about `flink run/run-applicaton` on the client side. For the JobManager and TaskManager start commands, you are right that `config.sh` will not be executed. So I think we could have a dedicated ticket to fix the issue you comment under [FLINK-24897](https://issues.apache.org/jira/browse/FLINK-24897?focusedCommentId=17483689&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17483689).

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Fixed in 986fff4543fbfd5016f087c53533b395af8c166f

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: No problem. In this pr, `ClusterEntrypointUtils.tryFindUserLibDirectory()` is used since 986fff4543fbfd5016f087c53533b395af8c166f.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Fix in 986fff4543fbfd5016f087c53533b395af8c166f

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java [hunk 8]
```diff
@@ -699,6 +700,41 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
         }
     }
 
+    /**
+     * Tests that the usrlib is added ship files again when local usrlib has been automatically
+     * shipped.
+     */
+    @Test
+    public void testShipUsrLibWithExistingLocalUsrLib() throws IOException {
```
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java]: Fixed in 986fff4543fbfd5016f087c53533b395af8c166f

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java]: It makes sense to me. I will update the JIRA as well to emphasize  we should always not include `usrlib` in the user shipped files.   Codes are much easier with this.
Fixed in 986fff4543fbfd5016f087c53533b395af8c166f

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java [hunk 9]
```diff
@@ -699,6 +699,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
         }
     }
 
+    /** Tests that the usrlib will be automatically shipped. */
+    @Test
+    public void testShipUsrLib() throws IOException {
+        final Map<String, String> oldEnv = System.getenv();
+        Map<String, String> env = new HashMap<>(1);
```
Comment (wangyang0918) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:706]: Please use `final` if possible.

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java [hunk 10]
```diff
@@ -699,6 +699,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
...
+        assertTrue(usrLibFolder.mkdirs());
+        File usrLibFile = new File(usrLibFolder, "usrLibFile.jar");
+        assertTrue(usrLibFile.createNewFile());
+        env.put(ConfigConstants.ENV_FLINK_LIB_DIR, libFolder.getAbsolutePath());
+        CommonTestUtils.setEnv(env);
+
+        try (YarnClusterDescriptor descriptor = createYarnClusterDescriptor()) {
+            Set<File> effectiveShipFiles = new HashSet<>();
+            descriptor.addUsrLibFolderToShipFiles(effectiveShipFiles);
+            Assert.assertFalse(effectiveShipFiles.contains(usrLibFile));
```
Comment (wangyang0918) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:721]: Use `assertThat` is always recommended.

```
assertThat(effectiveShipFiles, containsInAnyOrder(usrLibFolder));
```

**Review by wangyang0918 — CHANGES_REQUESTED** [review]
Thanks for addressing the comments. It looks almost good enough and I just left some minor suggestions.

One more thing about the config option `yarn.per-job-cluster.include-user-jar`. Could we 
 rename it to `yarn.classpath.include-user-jar` and deprecate the old one?

**Comment by bgeng777** [comment]
Hi @wangyang0918 thanks again for the detailed review. Above comments has been resolved in latest commit .
For the renaming of config option `yarn.per-job-cluster.include-user-jar`, I have adopted your advice, added `withDeprecatedKeys` and updated comments in the code.
 But I want to check if I should update markdown docs in this PR as well. It is not much work though I checked some Flink PRs and it seems docs tend to be modified in a doc revision PR. Any guidance on which one is recommended is appreciated.

**Comment by wangyang0918** [comment]
We already have a PR #18450 to document how to use `usrlib`. But you still need to regenerate the config docs.
```
mvn package -Dgenerate-config-docs -pl flink-docs -am -nsu -DskipTests
```

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 11]
```diff
@@ -876,7 +874,6 @@ private ApplicationReport startAppMaster(
         if (providedLibDirs == null || providedLibDirs.isEmpty()) {
             addLibFoldersToShipFiles(systemShipFiles);
         }
-
```
Comment (wangyang0918) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:879]: Unnecessary change.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 12]
```diff
@@ -918,6 +915,18 @@ private ApplicationReport startAppMaster(
                                 : Path.CUR_DIR,
                         LocalResourceType.FILE);
 
+        // usrlib will be automatically shipped if it exists.
+        final Set<File> usrLibShipFiles = new HashSet<>();
```
Comment (wangyang0918) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:919]: Hmm. Maybe I did not make myself clear. I do not mean to remove the condition check.

```
if (ClusterEntrypointUtils.tryFindUserLibDirectory().isPresent()) {
    ... ...
}
```

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnApplicationFileUploaderTest.java [hunk 13]
```diff
@@ -100,6 +105,35 @@ public void testRegisterProvidedLocalResourcesWithDuplication() throws IOExcepti
...
+
+        assertThatThrownBy(
+                        () ->
+                                YarnApplicationFileUploader.from(
+                                        FileSystem.get(new YarnConfiguration()),
+                                        new Path(flinkHomeDir.getPath()),
+                                        providedLibDirs,
+                                        ApplicationId.newInstance(0, 0),
+                                        DFSConfigKeys.DFS_REPLICATION_DEFAULT))
+                .isInstanceOf(RuntimeException.class)
```
Comment (wangyang0918) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnApplicationFileUploaderTest.java:130]: Maybe we could use a more specific `IllegalArgumentException` here.

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java [hunk 14]
```diff
@@ -699,6 +701,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
...
+        File usrLibFile = new File(usrLibFolder, "usrLibFile.jar");
+        assertTrue(usrLibFile.createNewFile());
+        env.put(ConfigConstants.ENV_FLINK_LIB_DIR, libFolder.getAbsolutePath());
+        CommonTestUtils.setEnv(env);
+
+        try (YarnClusterDescriptor descriptor = createYarnClusterDescriptor()) {
+            Set<File> effectiveShipFiles = new HashSet<>();
+            descriptor.addUsrLibFolderToShipFiles(effectiveShipFiles);
+            assertThat(effectiveShipFiles, containsInAnyOrder(usrLibFolder));
+            assertThat(effectiveShipFiles, not(containsInAnyOrder(usrLibFile)));
```
Comment (wangyang0918) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:724]: I do not think we still need this assert. The above `containsInAnyOrder` could cover such case.

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java [hunk 15]
```diff
@@ -699,6 +701,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
...
+        assertTrue(usrLibFolder.mkdirs());
+        File usrLibFile = new File(usrLibFolder, "usrLibFile.jar");
+        assertTrue(usrLibFile.createNewFile());
+        env.put(ConfigConstants.ENV_FLINK_LIB_DIR, libFolder.getAbsolutePath());
+        CommonTestUtils.setEnv(env);
+
+        try (YarnClusterDescriptor descriptor = createYarnClusterDescriptor()) {
+            Set<File> effectiveShipFiles = new HashSet<>();
+            descriptor.addUsrLibFolderToShipFiles(effectiveShipFiles);
+            assertThat(effectiveShipFiles, containsInAnyOrder(usrLibFolder));
```
Comment (wangyang0918) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:723]: Please use `org.hamcrest.MatcherAssert.assertThat` instead.

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java [hunk 16]
```diff
@@ -699,6 +701,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
         }
     }
 
+    /** Tests that the usrlib will be automatically shipped. */
+    @Test
+    public void testShipUsrLib() throws IOException {
+        final Map<String, String> oldEnv = System.getenv();
+        final Map<String, String> env = new HashMap<>(1);
+        File homeFolder = temporaryFolder.newFolder().getAbsoluteFile();
```
Comment (wangyang0918) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:709]: I believe `homeFolder` `libFolder` `usrLibFolder` `usrLibFile` `effectiveShipFiles` could also be final.

**Review by wangyang0918 — CHANGES_REQUESTED** [review]

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java (hunk 14 above)
Comment (bgeng777) [reply flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java]: The `assertThat(effectiveShipFiles, not(containsInAnyOrder(usrLibFile)));` just follows the check logic in testEnvironmentDirectoryShipping to make sure single is not added.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 12 above)
Comment (bgeng777) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: I remove this check because I double checked the codes and it should work well as we will do this check in addUsrLibFolderToShipFiles. 
For readability and less object allocation/function call , I agree to get it back.

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java (hunk 14 above)
Comment (bgeng777) [reply flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java]: You are right. Our `tryFindUserLibDirectory` will only return a directory.

**Review by wangyang0918 — APPROVED** [review]
LTGM. +1

I will rebase master, squash the commits and then merge this PR once the azure pipeline gives a pass.

**Comment by wangyang0918** [comment]
@flinkbot run azure

**Comment by bgeng777** [comment]
The azure test pipeline is broken for now and no e2e tests can succeed for all tests for now. Will run azure test tomorrow if it works.

**Comment by bgeng777** [comment]
Rerun the Flink CI in my own Azure pipeline and get a pass:
https://dev.azure.com/samuelgeng7/Flink/_build/results?buildId=74&view=results

**Comment by wangyang0918** [comment]
Merging this PR now.
